### ISSUE TITLE
Providing mechanism to retrieve params with values? 

### ISSUE THREAD

**Issue by stephenhandley** [pillarjs/path-to-regexp#153]
I think it would be helpful to have a convenience method `parseParams(path)` that provides a means of obtaining an object mapping from key names to their actual values rather than needing library consumers to do the following like express does. seems like a pretty common use-case so maybe I'm missing something in the docs.. surprised it isn't provided.
```
  this.params = {};
  this.path = match[0]

  var keys = this.keys;
  var params = this.params;

  for (var i = 1; i < match.length; i++) {
    var key = keys[i - 1];
    var prop = key.name;
    var val = decode_param(match[i])

    if (val !== undefined || !(hasOwnProperty.call(params, prop))) {
      params[prop] = val;
    }
  }
```
https://github.com/expressjs/express/blob/9722202df964bfbfc0f579e4baeb5a4e1b43b344/lib/router/layer.js#L139-L153

**Comment by dougwilson** [comment]
I think what you're describing is this library: https://github.com/pillarjs/path-match

**Comment by stephenhandley** [comment]
Ahah, figured I was missing something. thx!

**Comment by stephenhandley** [comment]
@dougwilson actually the issue here is that this lib obfuscates the underlying vars in your `path-to-regexp` lib.. 
https://github.com/pillarjs/path-match/blob/master/index.js#L9-L10

i.e doesn't expose the keys which I need for my use case (matching a route based on name and params). given that this implementation is so short, what's the rationale for it being in a separate library? looks like express is basically implementing a variant of what path-match does.

**Comment by dougwilson** [comment]
I'm not familiar enough with the histories to say, so would rather let the author of said modules chime in :)

**Comment by blakeembrey** [comment]
Seems reasonable to pull the raw logic into here nowadays. What's the expected signature? Example:

```ts
(re: RegExp, keys: Keys) => { [key: string]: string | string[] }
```

**Comment by damianobarbati** [comment]
+1 for this! why another lib?

**Comment by myateznik** [comment]
I think that you could use RegExp named capture groups. Example:

```JavaScript
let expression = pathToRegExp('/:foo/:bar')

let result = expression.exec('/hello/world')
// ['/hello/world', 'hello', 'world']

result.groups
// { foo: 'hello', bar: 'world' }
```

The only thing is that it will be supported in Node.js >= 10.3.0 (without --harmony flag) and Node.js >= 8.6.0 (with --harmony flag).

It seems to me that this is the best way, if we are guided by the latest LTS and stable versions of Node.js

**Comment by lirbank** [comment]
@myateznik that's a great idea! Havving an option for returning a RegExp with named capture groups for those using the latest JS (eg TypeScript, Babel, solves this for most projects anyways) would be awesome.

Eg something like:

```js
// Vanilla regex (today)

const re = pathToRegexp('/foo/:bar'})
// re = /^\/foo\/([^\/]+?)\/?$/i

const result = re.exec('/foo/123');
// result.groups = undefined
```

```js
// Regex with named capture groups (future)

const re = pathToRegexp('/foo/:bar', _, {namedCaptureGroups: true})
// re = /^\/foo\/(?<bar>[^\/]+?)\/?$/i

const result = re.exec('/foo/123');
// result.groups = {bar: "123"}
```

**Comment by lirbank** [comment]
@blakeembrey I was reading through #159 - I think a function that returns the params as key/value would work just as well as adding support for named capture groups. Actually, a function would probably work better than named capture groups for most users as it would require less boilerplate, assuming most users probably just want to get the params from a url as a map (object).

TLDR - I vote for a function per https://github.com/pillarjs/path-to-regexp/issues/159#issuecomment-406901917

**Comment by lirbank** [comment]
Simple function for getting the params from a URL as a map:

```js
function getParams(path: pathToRegexp.Path): { [key: string]: string } {
  const keys: pathToRegexp.Key[] = [];
  const r = pathToRegexp(path, keys).exec(location.pathname) || [];

  return keys.reduce((acc, key, i) => ({ [key.name]: r[i + 1], ...acc }), {});
}
```

**Comment by gilbert** [comment]
Any updates on this feature? Is it on the roadmap?

**Comment by blakeembrey** [comment]
@gilbert This module doesn't have a roadmap, but does accept PRs 😄

**Comment by blakeembrey** [comment]
Implemented with https://github.com/pillarjs/path-to-regexp/commit/0e0dce99b91b3deba85c2b5e97b161762a13c5d9.
